webserver_test.py

```diff
@@ -85,58 +71,61 @@
-sys.stdout.write("==========\n\n")
-for string in curl_output_expected_static_image:
-    pattern = re.compile(string)
-    if not pattern.search(curl_output):
-        ec = 1
-        sys.stdout.write("FAILED to match the following regular expression:\n")
-        sys.stdout.write("  " + string + "\n")
+ec = outputChecker(curl, curl_output_expected_static_image)
 
 # Request a file from the webserver using curl that does not exist
 curl = Popen(["curl", "-0", "-s", "localhost:2020/doesnotexist"], stdout=PIPE)
 curl_output = curl.communicate()[0].decode()
-sys.stdout.write("\nServer output for integration test 5 - handler dne\n")
+sys.stdout.write("\nServer output for integration test\n")
 sys.stdout.write("==========\n")
 sys.stdout.write(curl_output)
 sys.stdout.write("\n==========\n\n")
 if not curl_output == "<html><head><title>Not Found</title></head><body><h1>404 Not Found</h1></body></html>":
     ec = 1
-    sys.stdout.write("FAILED to match the following expression:\n")
+    sys.stdout.write("FAILED to match the following regular expression:\n")
     sys.stdout.write("  <html><head><title>Not Found</title></head><body><h1>404 Not Found</h1></body></html>\n")
 
-# Test the multithreading capabilities of the webserver
-s1 = socket.socket()
-s1.connect(("localhost", 2020))
-s1.send("GET /echo HT")
-s2 = socket.socket()
-s2.connect(("localhost", 2020))
-s2.send("GET /echo HTTP/1.0\r\n\r\n")
-s2.settimeout(1)
-multithreaded_output = s2.recv(1024)
-s1.close()
-s2.close()
-sys.stdout.write("\nServer output for integration test 6 - multithreading\n")
+# Request echo response from the  reverse proxy webserver
+curl = Popen(["curl", "-0", "-s", "localhost:4242/reverse_proxy/echo"], stdout=PIPE)
+ec = outputChecker(curl, curl_output_expected_echo)
+
+# Request a file from the  reverse proxy webserver
+curl = Popen(["curl", "-0", "-s", "localhost:4242/reverse_proxy/test_file"], stdout=PIPE)
+curl_output = curl.communicate()[0].decode()
+sys.stdout.write("\nServer output for integration test\n")
 sys.stdout.write("==========\n")
-sys.stdout.write(multithreaded_output)
-sys.stdout.write("\n==========\n\n")
-if not multithreaded_output == expected_multithreaded_output:
+sys.stdout.write(curl_output)
+sys.stdout.write("==========\n\n")
+if not curl_output == "TEST\n":
+    ec = 1
+    sys.stdout.write("FAILED to match the following regular expression:\n")
+    sys.stdout.write("  TEST\n")
+
+# Request an image from the  reverse proxy webserver
+curl = Popen(["curl", "-0", "-s", "-o", "proxy_bunny", "localhost:4242/reverse_proxy/bunny.jpg"], stdout=PIPE)
+diff = Popen(["cmp", "bunny.jpg", "proxy_bunny"], stdout=PIPE)
+if (diff.communicate()[0].decode() != ""):
+    ec = 1
+    sys.stdout.write("FAILED to download matching copy of bunny.jpg via reverse-proxy\n")
+
+# Request ucla.edu from reverse proxy webserver for 302 testing
+curl = Popen(["curl", "-0", "-s", "localhost:4242/redirect"], stdout=PIPE)
+curl_output = curl.communicate()[0].decode("utf-8")
```

Comment:
UTF-8 encoding/decoding is required because of non-ASCII characters on the UCLA website, likely accents in names, for example.